### PR TITLE
[Merged by Bors] - feat: category of coalgebras

### PR DESCRIPTION
**Pull request by 101damnations**
The category of coalgebras over a commutative ring. Mimics `Mathlib.LinearAlgebra.QuadraticForm.QuadraticModuleCat`.

---

- [x] depends on: #11961 
- [x] depends on: #11966 
- [x] depends on: #11970 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by eric-wieser — COMMENTED** [review]
As this is new code, I wonder if we should copy the strategy of the `QuadraticModuleCat` instead, which has slightly more opaque types and is probably more performant as a result.

cc @mattrobball, @joelriou

**Review by kbuzzard — COMMENTED** [review]
I don't understand Eric's comment yet but this all looks good to me.

File: Mathlib/Algebra/Category/CoalgebraCat/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  rfl
+
+end CategoryTheory.Iso
+
+/-- Coalgebra equivalences are the same as (isomorphic to) isomorphisms
+in `CoalgebraCat R` -/
+@[simps]
+def coalgEquivIsoCoalgebraIso {X Y : Type u} [AddCommGroup X] [AddCommGroup Y]
+    [Module R X] [Coalgebra R X] [Module R Y] [Coalgebra R Y] :
+    (X ≃ₗc[R] Y) ≅ CoalgebraCat.of R X ≅ CoalgebraCat.of R Y where
```
[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/Algebra/Category/CoalgebraCat/Basic.lean]: Thanks, that was a mistake

**Comment by 101damnations** [comment]
Thanks for the suggestion @eric-wieser - I have tried to apply it

File: Mathlib/Algebra/Category/CoalgebraCat/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+@[simp] theorem toCoalgHom_id {M : CoalgebraCat.{v} R} :
+    Hom.toCoalgHom (𝟙 M) = CoalgHom.id _ _ :=
+  rfl
+
+instance concreteCategory : ConcreteCategory.{v} (CoalgebraCat.{v} R) where
+  forget :=
+    { obj := fun M => M
+      map := @fun M N f => f.toCoalgHom }
+  forget_faithful :=
+    { map_injective := @fun M N => DFunLike.coe_injective.comp <| Hom.toCoalgHom_injective _ _ }
```
Comment (kbuzzard) [inline Mathlib/Algebra/Category/CoalgebraCat/Basic.lean:103]:
```suggestion
      map := fun {M N} f => f.toCoalgHom }
  forget_faithful :=
    { map_injective := fun {M N} => DFunLike.coe_injective.comp <| Hom.toCoalgHom_injective _ _ }
```
This way avoids the `@`s

**Review by kbuzzard — COMMENTED** [review]
@eric-wieser are you happier with this approach?

**Review by eric-wieser — APPROVED** [review]
This looks good to me, thanks! I'll ping @joelriou one more time in case they have any post-merge opinions, but I think any objections are probably worth a larger discussion + refactor anyway.

bors merge

**Comment by eric-wieser** [comment]
> I don't understand Eric's comment yet but this all looks good to me.

The key thing is that we change the category-theoretic `Hom` to be a one-field wrapper for `CoAlgebraHom`, which means that you get the explicitly-typed coercion functions for free and can state lemmas more clearly with them.

**Pull request closed without merging**